### PR TITLE
Implemented round robin and other spec related features for ReqSocket

### PR DESCRIPTION
**Pull request by Alexei-Kornienko**
Now this socket seems to be spec compliant. Need to test it properly and benchmark overall code

### CODE REVIEW

File: src/req_rep.rs [hunk 1]
```diff
@@ -119,6 +151,7 @@ impl MultiPeer for ReqSocketBackend {
                 _io_close_handle: stop_handle,
```
Comment (malor) [inline src/req_rep.rs:151]: Hmm, why do we need to synchronise access to `in_queue_receiver`? Isn't the receiving end of this channel only used in `ReqSocket::recv()`?

File: src/req_rep.rs [hunk 2]
```diff
@@ -119,6 +151,7 @@ impl MultiPeer for ReqSocketBackend {
...
             },
```
Comment (malor) [inline src/req_rep.rs:152]: Don't we have at most 1 outstanding request in ZMQ_REQ sockets? Is there a reason to have channels with size > 1?

File: src/req_rep.rs [hunk 3]
```diff
@@ -25,21 +28,43 @@ pub struct ReqSocket {
...
+                    let frames = vec![
+                        "".into(), // delimiter frame
+                        message,
+                    ];
+                    peer.send_queue
+                        .send(Message::MultipartMessage(frames))
+                        .await?;
+                    self.backend
+                        .current_request_peer_id
+                        .store(self.backend.peers.hash_usize(&next_peer_id));
```
Comment (malor) [inline src/req_rep.rs:61]: This should be fine, but `hash_usize()` is *probably* not on par with cryptographic hash functions like SHA-1, so I'd expect the probability of collisions to be higher. IIUC, the only reason you are going for using of the hash function in the first place is that you want to use an atomic, but peer identity is a vector of u8's and can't be used as one. Perhaps, an alternative would be to map peer identity to an incrementing integer and use that instead.

File: src/req_rep.rs [hunk 4]
```diff
@@ -78,14 +103,21 @@ impl SocketFrontend for ReqSocket {
...
+                peers: DashMap::new(),
+                round_robin: SegQueue::new(),
+                current_request_peer_id: AtomicCell::new(0),
             }),
             _accept_close_handle: None,
             current_request: None,
         }
     }
 
     async fn bind(&mut self, endpoint: &str) -> ZmqResult<()> {
```
Comment (malor) [inline src/req_rep.rs:115]: Hmm, shouldn't a ZMQ_REQ socket only have `connect()`, and not `bind()`?

Comment (Alexei-Kornienko) [reply src/req_rep.rs:115]: No all zmq sockets can work both as client or server. It depends on actual use case which variant is used

**Comment by Alexei-Kornienko** [comment]
Agreed on most things on a call. Possibly will return to some remaining issues later in another request

**Pull request merged**